enterprise/server/scheduling/scheduler_server/scheduler_server.go

```diff
@@ -1671,110 +1662,76 @@ func (s *SchedulerServer) enqueueTaskReservations(ctx context.Context, enqueueRe
 	// Note: preferredNode may be nil if the executor ID isn't specified or if
 	// the executor is no longer connected.
 	preferredNode := nodeBalancer.FindConnectedExecutorByID(enqueueRequest.GetExecutorId())
+	if preferredNode != nil {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			break
+		}
+		enqueueStart := time.Now()
+		if s.enqueue(ctx, preferredNode, enqueueRequest, opts) {
+			scheduledOnPreferredNode = true
+			successfulReservations = append(successfulReservations, successfulReservation(preferredNode, enqueueStart))
+		}
+	}
 
 	attempts := 0
-	var rankedNodes []*rankedExecutionNode
-	sampleIndex := 0
+	var rankedNodes []interfaces.RankedExecutionNode
 	nonPreferredDelay := getNonPreferredSchedulingDelay(cmd)
-	delaySpecified := enqueueRequest.GetDelay() != nil
-	for probesSent < probeCount {
-		attempts++
+	delayable := enqueueRequest.GetDelay() == nil
+	for len(successfulReservations) < probeCount {
```

Comment:
Since `successfulReservations` is a list of "debug strings" it feels slightly unsafe to be using it for control flow here.

I would prefer to bring back the `probesSent` counter and then do the control flow based on that, then keep `successfulReservations` purely for logging/debugging purposes

Reply:
I'd kind of like to not have a variable that's just keeping track of the length of this slice, because keeping track of 2 variables separately is more bug-prone than just 1. If you'd prefer, I could introduce a little struct for tracking the successful reservations with a `.add()` function, but I'd prefer to just rely on the length of the slice.